<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (version 1.7.0_21) on Tue Oct 29 14:06:36 GMT 2013 -->
<meta http-equiv="Content-Type" content="text/html" charset="UTF-8">
<title>G-Index (GenomeHelper)</title>
<meta name="date" content="2013-10-29">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
</head>
<body>
<script type="text/javascript"><!--
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="G-Index (GenomeHelper)";
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar_top">
<!--   -->
</a><a href="#skip-navbar_top" title="Skip navigation links"></a><a name="navbar_top_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-4.html">Prev Letter</a></li>
<li><a href="index-6.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-filesindex-5.html" target="_top">Frames</a></li>
<li><a href="index-5.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip-navbar_top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="contentContainer"><a href="index-1.html">C</a>&nbsp;<a href="index-2.html">D</a>&nbsp;<a href="index-3.html">E</a>&nbsp;<a href="index-4.html">F</a>&nbsp;<a href="index-5.html">G</a>&nbsp;<a href="index-6.html">H</a>&nbsp;<a href="index-7.html">I</a>&nbsp;<a href="index-8.html">J</a>&nbsp;<a href="index-9.html">M</a>&nbsp;<a href="index-10.html">O</a>&nbsp;<a href="index-11.html">P</a>&nbsp;<a href="index-12.html">Q</a>&nbsp;<a href="index-13.html">R</a>&nbsp;<a href="index-14.html">S</a>&nbsp;<a href="index-15.html">T</a>&nbsp;<a href="index-16.html">U</a>&nbsp;<a href="index-17.html">V</a>&nbsp;<a href="index-18.html">W</a>&nbsp;<a name="_G_">
<!--   -->
</a>
<h2 class="title">G</h2>
<dl>
<dt><a href="../uk/ac/tsl/etherington/genomehelper/main/GenomeHelper.html" title="class in uk.ac.tsl.etherington.genomehelper.main"><span class="strong">GenomeHelper</span></a> - Class in <a href="../uk/ac/tsl/etherington/genomehelper/main/package-summary.html">uk.ac.tsl.etherington.genomehelper.main</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/main/GenomeHelper.html#GenomeHelper()">GenomeHelper()</a></span> - Constructor for class uk.ac.tsl.etherington.genomehelper.main.<a href="../uk/ac/tsl/etherington/genomehelper/main/GenomeHelper.html" title="class in uk.ac.tsl.etherington.genomehelper.main">GenomeHelper</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fasta/FastaSubstrings.html#getAllSubStrings(java.lang.String)">getAllSubStrings(String)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.fasta.<a href="../uk/ac/tsl/etherington/genomehelper/fasta/FastaSubstrings.html" title="class in uk.ac.tsl.etherington.genomehelper.fasta">FastaSubstrings</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#getBadSeqFastqWriter(java.io.File, net.sf.picard.fastq.FastqWriterFactory)">getBadSeqFastqWriter(File, FastqWriterFactory)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.fastq.<a href="../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html" title="class in uk.ac.tsl.etherington.genomehelper.fastq">FastqQC</a></dt>
<dd>
<div class="block">Creates a FastqWriter for bad reads that fail Quality Control</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getBlocks(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String)">getBlocks(FeatureList, String)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.gff.<a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff">GFFFeatureStats</a></dt>
<dd>
<div class="block">Creates start/end blocks of features</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getBlocks(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, java.lang.String)">getBlocks(FeatureList, String, String)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.gff.<a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff">GFFFeatureStats</a></dt>
<dd>
<div class="block">Creates start/end blocks of features</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getBlocks(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, java.lang.String, java.util.Set)">getBlocks(FeatureList, String, String, Set)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.gff.<a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff">GFFFeatureStats</a></dt>
<dd>
<div class="block">Creates start/end blocks of features</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getBothPairedMappedSamRecords(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File)">getBothPairedMappedSamRecords(File, File, File, File, File)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.bam.<a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html" title="class in uk.ac.tsl.etherington.genomehelper.bam">MappedSamRecords</a></dt>
<dd>
<div class="block">Returns the pairs of sequences where at both of the pair has been mapped
 to a genome.</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getBothPairedUnmappedSamRecords(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File)">getBothPairedUnmappedSamRecords(File, File, File, File, File)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.bam.<a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html" title="class in uk.ac.tsl.etherington.genomehelper.bam">MappedSamRecords</a></dt>
<dd>
<div class="block">Returns the pairs of sequences where both of the pair has not been mapped
 to a genome.</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fasta/ContigPosition.html#getContigid()">getContigid()</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.fasta.<a href="../uk/ac/tsl/etherington/genomehelper/fasta/ContigPosition.html" title="class in uk.ac.tsl.etherington.genomehelper.fasta">ContigPosition</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fasta/ContigPosition.html#getEnd()">getEnd()</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.fasta.<a href="../uk/ac/tsl/etherington/genomehelper/fasta/ContigPosition.html" title="class in uk.ac.tsl.etherington.genomehelper.fasta">ContigPosition</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getFeatureList(java.lang.String)">getFeatureList(String)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.gff.<a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff">GFFFeatureStats</a></dt>
<dd>
<div class="block">Gets all the features from a gff file</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fasta/FastaFeatures.html#getGCContent(java.io.File)">getGCContent(File)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.fasta.<a href="../uk/ac/tsl/etherington/genomehelper/fasta/FastaFeatures.html" title="class in uk.ac.tsl.etherington.genomehelper.fasta">FastaFeatures</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fasta/FastaFeatures.html#getGenomeSize(java.io.File)">getGenomeSize(File)</a></span> - Static method in class uk.ac.tsl.etherington.genomehelper.fasta.<a href="../uk/ac/tsl/etherington/genomehelper/fasta/FastaFeatures.html" title="class in uk.ac.tsl.etherington.genomehelper.fasta">FastaFeatures</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getGenomeSizeFromIntArrayHashMap(java.util.HashMap)">getGenomeSizeFromIntArrayHashMap(HashMap&lt;String, int[]&gt;)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.gff.<a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff">GFFFeatureStats</a></dt>
<dd>
<div class="block">Calculates the length of a genome from collection of int arrays, where
 each int array is the length of a genome sequence</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getMeanFeatureLength(org.biojava3.genome.parsers.gff.FeatureList, java.util.HashMap, java.lang.String)">getMeanFeatureLength(FeatureList, HashMap&lt;String, int[]&gt;, String)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.gff.<a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff">GFFFeatureStats</a></dt>
<dd>
<div class="block">Prints stats and mean feature length for a given feature</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getMeanFeatureLength(org.biojava3.genome.parsers.gff.FeatureList, java.util.HashMap, java.lang.String, java.io.File, java.lang.String)">getMeanFeatureLength(FeatureList, HashMap&lt;String, int[]&gt;, String, File, String)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.gff.<a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff">GFFFeatureStats</a></dt>
<dd>
<div class="block">Prints stats and mean feature length for a given feature, but restricts
 it to a list of geneIds of interest</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getMeanFeatureLength(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String)">getMeanFeatureLength(FeatureList, String)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.gff.<a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff">GFFFeatureStats</a></dt>
<dd>
<div class="block">Calculates the mean feature length of a given feature</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getMeanIntronLength(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, double)">getMeanIntronLength(FeatureList, String, double)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.gff.<a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff">GFFFeatureStats</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getMeanSecretedIntronLength(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, java.io.File)">getMeanSecretedIntronLength(FeatureList, String, File)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.gff.<a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff">GFFFeatureStats</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getOnePairedMappedSamRecords(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File)">getOnePairedMappedSamRecords(File, File, File, File, File)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.bam.<a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html" title="class in uk.ac.tsl.etherington.genomehelper.bam">MappedSamRecords</a></dt>
<dd>
<div class="block">Returns the pairs of sequences where at least one of the pair has been
 mapped to a genome.</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getOnePairedUnmappedSamRecords(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File)">getOnePairedUnmappedSamRecords(File, File, File, File, File)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.bam.<a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html" title="class in uk.ac.tsl.etherington.genomehelper.bam">MappedSamRecords</a></dt>
<dd>
<div class="block">Returns the pairs of sequences where at least one of the pair has not
 been mapped to a genome.</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html#getOneSideFastqSeqsFromList(java.io.File, java.io.File, java.io.File)">getOneSideFastqSeqsFromList(File, File, File)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.fastq.<a href="../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html" title="class in uk.ac.tsl.etherington.genomehelper.fastq">FastqParser</a></dt>
<dd>
<div class="block">Extract fastq sequences in a list from a file of single-end fastq
 sequences</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html#getPairedFastqSeqsFromHashSet(java.util.HashSet, java.io.File, java.io.File, java.io.File, java.io.File)">getPairedFastqSeqsFromHashSet(HashSet, File, File, File, File)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.fastq.<a href="../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html" title="class in uk.ac.tsl.etherington.genomehelper.fastq">FastqParser</a></dt>
<dd>
<div class="block">Extract paired-end fastq sequences from a list of sequence names</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fasta/FastaFeatures.html#getParsedDNASequences(java.io.File)">getParsedDNASequences(File)</a></span> - Static method in class uk.ac.tsl.etherington.genomehelper.fasta.<a href="../uk/ac/tsl/etherington/genomehelper/fasta/FastaFeatures.html" title="class in uk.ac.tsl.etherington.genomehelper.fasta">FastaFeatures</a></dt>
<dd>
<div class="block">Parses a multi-fasta file into a LinkedHashMap of sequence names and DNASequences</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fastq/FastqMotifFinder.html#getPEFastqReadsFromMotif(java.io.File, java.lang.String, java.io.File)">getPEFastqReadsFromMotif(File, String, File)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.fastq.<a href="../uk/ac/tsl/etherington/genomehelper/fastq/FastqMotifFinder.html" title="class in uk.ac.tsl.etherington.genomehelper.fastq">FastqMotifFinder</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getReads(net.sf.samtools.SAMFileReader, net.sf.samtools.SAMRecordIterator, boolean)">getReads(SAMFileReader, SAMRecordIterator, boolean)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.bam.<a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html" title="class in uk.ac.tsl.etherington.genomehelper.bam">MappedSamRecords</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fasta/FastaFeatures.html#getSequenceAsIntArray(java.io.File)">getSequenceAsIntArray(File)</a></span> - Static method in class uk.ac.tsl.etherington.genomehelper.fasta.<a href="../uk/ac/tsl/etherington/genomehelper/fasta/FastaFeatures.html" title="class in uk.ac.tsl.etherington.genomehelper.fasta">FastaFeatures</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fasta/FastaFeatures.html#getSequenceLengths(java.io.File)">getSequenceLengths(File)</a></span> - Static method in class uk.ac.tsl.etherington.genomehelper.fasta.<a href="../uk/ac/tsl/etherington/genomehelper/fasta/FastaFeatures.html" title="class in uk.ac.tsl.etherington.genomehelper.fasta">FastaFeatures</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getSingleMappedSamRecords(java.io.File, java.io.File, java.io.File, boolean)">getSingleMappedSamRecords(File, File, File, boolean)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.bam.<a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html" title="class in uk.ac.tsl.etherington.genomehelper.bam">MappedSamRecords</a></dt>
<dd>
<div class="block">Returns the fastq sequences that have been mapped to a genome.</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getSingleUnmappedSamRecords(java.io.File, java.io.File, java.io.File, boolean)">getSingleUnmappedSamRecords(File, File, File, boolean)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.bam.<a href="../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html" title="class in uk.ac.tsl.etherington.genomehelper.bam">MappedSamRecords</a></dt>
<dd>
<div class="block">Returns the fastq sequences that have not been mapped to a genome.</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fasta/ContigPosition.html#getStart()">getStart()</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.fasta.<a href="../uk/ac/tsl/etherington/genomehelper/fasta/ContigPosition.html" title="class in uk.ac.tsl.etherington.genomehelper.fasta">ContigPosition</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getStats(java.lang.String, java.io.File, java.lang.String)">getStats(String, File, String)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.gff.<a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff">GFFFeatureStats</a></dt>
<dd>
<div class="block">Combines a number of methods</div>
</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fasta/ContigPosition.html#getStrand()">getStrand()</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.fasta.<a href="../uk/ac/tsl/etherington/genomehelper/fasta/ContigPosition.html" title="class in uk.ac.tsl.etherington.genomehelper.fasta">ContigPosition</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff"><span class="strong">GFFFeatureStats</span></a> - Class in <a href="../uk/ac/tsl/etherington/genomehelper/gff/package-summary.html">uk.ac.tsl.etherington.genomehelper.gff</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#GFFFeatureStats()">GFFFeatureStats()</a></span> - Constructor for class uk.ac.tsl.etherington.genomehelper.gff.<a href="../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" title="class in uk.ac.tsl.etherington.genomehelper.gff">GFFFeatureStats</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#groomRead(net.sf.picard.fastq.FastqRecord, java.lang.String)">groomRead(FastqRecord, String)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.fastq.<a href="../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html" title="class in uk.ac.tsl.etherington.genomehelper.fastq">FastqQC</a></dt>
<dd>&nbsp;</dd>
<dt><span class="strong"><a href="../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#guessFormat(net.sf.picard.fastq.FastqRecord)">guessFormat(FastqRecord)</a></span> - Method in class uk.ac.tsl.etherington.genomehelper.fastq.<a href="../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html" title="class in uk.ac.tsl.etherington.genomehelper.fastq">FastqQC</a></dt>
<dd>
<div class="block">Guesses the format of a FastqRecord</div>
</dd>
</dl>
<a href="index-1.html">C</a>&nbsp;<a href="index-2.html">D</a>&nbsp;<a href="index-3.html">E</a>&nbsp;<a href="index-4.html">F</a>&nbsp;<a href="index-5.html">G</a>&nbsp;<a href="index-6.html">H</a>&nbsp;<a href="index-7.html">I</a>&nbsp;<a href="index-8.html">J</a>&nbsp;<a href="index-9.html">M</a>&nbsp;<a href="index-10.html">O</a>&nbsp;<a href="index-11.html">P</a>&nbsp;<a href="index-12.html">Q</a>&nbsp;<a href="index-13.html">R</a>&nbsp;<a href="index-14.html">S</a>&nbsp;<a href="index-15.html">T</a>&nbsp;<a href="index-16.html">U</a>&nbsp;<a href="index-17.html">V</a>&nbsp;<a href="index-18.html">W</a>&nbsp;</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar_bottom">
<!--   -->
</a><a href="#skip-navbar_bottom" title="Skip navigation links"></a><a name="navbar_bottom_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-4.html">Prev Letter</a></li>
<li><a href="index-6.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-filesindex-5.html" target="_top">Frames</a></li>
<li><a href="index-5.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip-navbar_bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
