<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (version 1.7.0_21) on Wed Dec 18 15:53:34 GMT 2013 -->
<meta http-equiv="Content-Type" content="text/html" charset="UTF-8">
<title>FastqQC (GenomeHelper)</title>
<meta name="date" content="2013-12-18">
<link rel="stylesheet" type="text/css" href="../../../../../../stylesheet.css" title="Style">
</head>
<body>
<script type="text/javascript"><!--
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="FastqQC (GenomeHelper)";
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar_top">
<!--   -->
</a><a href="#skip-navbar_top" title="Skip navigation links"></a><a name="navbar_top_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/FastqQC.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html" title="class in uk.ac.tsl.etherington.genomehelper.fastq"><span class="strong">Prev Class</span></a></li>
<li>Next Class</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html" target="_top">Frames</a></li>
<li><a href="FastqQC.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">uk.ac.tsl.etherington.genomehelper.fastq</div>
<h2 title="Class FastqQC" class="title">Class FastqQC</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>uk.ac.tsl.etherington.genomehelper.fastq.FastqQC</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<hr>
<br>
<pre>public class <span class="strong">FastqQC</span>
extends java.lang.Object</pre>
<div class="block">A Quality Control (QC) class that takes fastq sequences in various structure
 and quality format. The sequences are checked for Ns, long/short reads and
 format. With paired-end reads, if either of the pairs fails QC then neither
 of the pair is returned. Reads with the quality format of 'illumina'
 (pre-Illumina 1.8) are automatically changed to sanger quality format.
 Currently only accepts Illumina 1.3+</div>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#FastqQC()">FastqQC</a></strong>()</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method_summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span>Methods</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#checkFormat(java.lang.String)">checkFormat</a></strong>(java.lang.String&nbsp;format)</code>
<div class="block">Checks that a valid format ('illumina' or 'sanger') has been supplied</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#countLengthsAndNs(java.io.File)">countLengthsAndNs</a></strong>(java.io.File&nbsp;fastq)</code>
<div class="block">Takes a fastq file and calculates the number of reads which have a 'N'
 and the distribution of read lengths</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>net.sf.picard.fastq.FastqWriter</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#getBadSeqFastqWriter(java.io.File, net.sf.picard.fastq.FastqWriterFactory)">getBadSeqFastqWriter</a></strong>(java.io.File&nbsp;fastqFileOut,
                    net.sf.picard.fastq.FastqWriterFactory&nbsp;writer)</code>
<div class="block">Creates a FastqWriter for bad reads that fail Quality Control</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>net.sf.picard.fastq.FastqRecord</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#groomRead(net.sf.picard.fastq.FastqRecord, java.lang.String)">groomRead</a></strong>(net.sf.picard.fastq.FastqRecord&nbsp;record,
         java.lang.String&nbsp;format)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>net.sf.picard.util.FastqQualityFormat</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#guessFormat(net.sf.picard.fastq.FastqRecord)">guessFormat</a></strong>(net.sf.picard.fastq.FastqRecord&nbsp;record)</code>
<div class="block">Guesses the format of a FastqRecord</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#main(java.lang.String[])">main</a></strong>(java.lang.String[]&nbsp;args)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#qcFastq(net.sf.picard.fastq.FastqRecord, int)">qcFastq</a></strong>(net.sf.picard.fastq.FastqRecord&nbsp;seqRecord,
       int&nbsp;singleEndReadLength)</code>
<div class="block">Quality controls a fastq read</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#qcInterlacedReads(java.io.File, java.io.File, int, java.lang.String, boolean)">qcInterlacedReads</a></strong>(java.io.File&nbsp;fastqFileIn,
                 java.io.File&nbsp;fastqFileOut,
                 int&nbsp;singleEndReadLength,
                 java.lang.String&nbsp;format,
                 boolean&nbsp;writeBadSeqs)</code>
<div class="block">Removes pairs of reads where at least one of the pair contains short/long
 reads or a read that contain an 'N', from an interlaced fastq file</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#qcInterlacedReadsToPairs(java.io.File, java.io.File, java.io.File, int, java.lang.String, boolean)">qcInterlacedReadsToPairs</a></strong>(java.io.File&nbsp;fastqFileIn,
                        java.io.File&nbsp;leftReadsOut,
                        java.io.File&nbsp;rightReadsOut,
                        int&nbsp;singleEndReadLength,
                        java.lang.String&nbsp;format,
                        boolean&nbsp;writeBadSeqs)</code>
<div class="block">Removes pairs of reads where at least one of the pair contains short/long
 reads or a read that contain an 'N', from an interlaced fastq file</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#qcJoinedReads(java.io.File, java.io.File, java.io.File, int, java.lang.String, boolean)">qcJoinedReads</a></strong>(java.io.File&nbsp;fastqFileIn,
             java.io.File&nbsp;leftReadsOut,
             java.io.File&nbsp;rightReadsOut,
             int&nbsp;readLength,
             java.lang.String&nbsp;format,
             boolean&nbsp;writeBadSeqs)</code>
<div class="block">Removes pairs of reads where at least one of the pair contains short/long
 reads or a read that contain an 'N', from a fastq file where the reads
 are joined</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#qcPairedReads(java.io.File, java.io.File, java.io.File, java.io.File, int, java.lang.String, boolean)">qcPairedReads</a></strong>(java.io.File&nbsp;leftFastqFileIn,
             java.io.File&nbsp;rightFastqFileIn,
             java.io.File&nbsp;leftReadsOut,
             java.io.File&nbsp;rightReadsOut,
             int&nbsp;singleEndReadLength,
             java.lang.String&nbsp;format,
             boolean&nbsp;writeBadSeqs)</code>
<div class="block">Removes pairs of reads where at least one of the pair contains short/long
 reads or a read that contain an 'N', from paired-end fastq files.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#qcSingleEndReads(java.io.File, java.io.File, int, java.lang.String, boolean)">qcSingleEndReads</a></strong>(java.io.File&nbsp;fastqFileIn,
                java.io.File&nbsp;readsOut,
                int&nbsp;singleEndReadLength,
                java.lang.String&nbsp;format,
                boolean&nbsp;writeBadSeqs)</code>
<div class="block">Removes reads that contain short/long reads or that contain an 'N', from
 a fastq file</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#veryfiyPairedReads(java.io.File, java.io.File)">veryfiyPairedReads</a></strong>(java.io.File&nbsp;fastqLeft,
                  java.io.File&nbsp;fastqRight)</code>
<div class="block">Verifies that two files of left/right paired-end reads are in order and
 contain no fastq format errors</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html#veryfiyReads(java.io.File)">veryfiyReads</a></strong>(java.io.File&nbsp;fastq)</code>
<div class="block">Checks a fastq file to verify that all the reads can be parsed into a
 FastqRecord.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods_inherited_from_class_java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="FastqQC()">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>FastqQC</h4>
<pre>public&nbsp;FastqQC()</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method_detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="checkFormat(java.lang.String)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>checkFormat</h4>
<pre>public&nbsp;boolean&nbsp;checkFormat(java.lang.String&nbsp;format)</pre>
<div class="block">Checks that a valid format ('illumina' or 'sanger') has been supplied</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>format</code> - the format (must only be 'illumina' or 'sanger')</dd>
<dt><span class="strong">Returns:</span></dt><dd>true of the format is valid, otherwise false</dd></dl>
</li>
</ul>
<a name="qcFastq(net.sf.picard.fastq.FastqRecord, int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>qcFastq</h4>
<pre>public&nbsp;boolean&nbsp;qcFastq(net.sf.picard.fastq.FastqRecord&nbsp;seqRecord,
              int&nbsp;singleEndReadLength)</pre>
<div class="block">Quality controls a fastq read</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>seqRecord</code> - a net.sf.picard.fastq.FastqRecord to QC</dd><dd><code>singleEndReadLength</code> - the expected length of a single end read</dd>
<dt><span class="strong">Returns:</span></dt><dd>true if the read is the correct length and contains no 'N's,
 false otherwise</dd></dl>
</li>
</ul>
<a name="getBadSeqFastqWriter(java.io.File, net.sf.picard.fastq.FastqWriterFactory)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getBadSeqFastqWriter</h4>
<pre>public&nbsp;net.sf.picard.fastq.FastqWriter&nbsp;getBadSeqFastqWriter(java.io.File&nbsp;fastqFileOut,
                                                   net.sf.picard.fastq.FastqWriterFactory&nbsp;writer)</pre>
<div class="block">Creates a FastqWriter for bad reads that fail Quality Control</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fastqFileOut</code> - a FastqWriter to write bad reads to</dd><dd><code>writer</code> - a FastqWriterFactory. Multiple FastqWriters can be
 generated from a single FastqWriterFactory</dd>
<dt><span class="strong">Returns:</span></dt><dd>a FastqWriter to write bad reads to. The bad reads file will
 start with 'bad_', followed by the outfile name.</dd></dl>
</li>
</ul>
<a name="qcPairedReads(java.io.File, java.io.File, java.io.File, java.io.File, int, java.lang.String, boolean)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>qcPairedReads</h4>
<pre>public&nbsp;void&nbsp;qcPairedReads(java.io.File&nbsp;leftFastqFileIn,
                 java.io.File&nbsp;rightFastqFileIn,
                 java.io.File&nbsp;leftReadsOut,
                 java.io.File&nbsp;rightReadsOut,
                 int&nbsp;singleEndReadLength,
                 java.lang.String&nbsp;format,
                 boolean&nbsp;writeBadSeqs)</pre>
<div class="block">Removes pairs of reads where at least one of the pair contains short/long
 reads or a read that contain an 'N', from paired-end fastq files.
 Optionally writes bad reads to a file.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>leftFastqFileIn</code> - the left-handed reads</dd><dd><code>rightFastqFileIn</code> - the right-handed reads</dd><dd><code>leftReadsOut</code> - the QCd left-handed reads</dd><dd><code>rightReadsOut</code> - the QCd right-handed reads</dd><dd><code>singleEndReadLength</code> - the expected length of a single read</dd><dd><code>format</code> - the fastq format (can only be 'illumina' or 'sanger')</dd><dd><code>writeBadSeqs</code> - whether to write the bad reads to a file (bad reads
 file name will start with 'bad_')</dd></dl>
</li>
</ul>
<a name="qcInterlacedReads(java.io.File, java.io.File, int, java.lang.String, boolean)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>qcInterlacedReads</h4>
<pre>public&nbsp;void&nbsp;qcInterlacedReads(java.io.File&nbsp;fastqFileIn,
                     java.io.File&nbsp;fastqFileOut,
                     int&nbsp;singleEndReadLength,
                     java.lang.String&nbsp;format,
                     boolean&nbsp;writeBadSeqs)</pre>
<div class="block">Removes pairs of reads where at least one of the pair contains short/long
 reads or a read that contain an 'N', from an interlaced fastq file</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fastqFileIn</code> - an interlaced fastq file</dd><dd><code>fastqFileOut</code> - the good left-handed reads</dd><dd><code>singleEndReadLength</code> - the expected length of a single read</dd><dd><code>format</code> - the fastq format (can only be 'illumina' or 'sanger')</dd><dd><code>writeBadSeqs</code> - whether to write the bad reads to a file (bad reads
 file name will start with 'bad_')</dd></dl>
</li>
</ul>
<a name="qcInterlacedReadsToPairs(java.io.File, java.io.File, java.io.File, int, java.lang.String, boolean)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>qcInterlacedReadsToPairs</h4>
<pre>public&nbsp;void&nbsp;qcInterlacedReadsToPairs(java.io.File&nbsp;fastqFileIn,
                            java.io.File&nbsp;leftReadsOut,
                            java.io.File&nbsp;rightReadsOut,
                            int&nbsp;singleEndReadLength,
                            java.lang.String&nbsp;format,
                            boolean&nbsp;writeBadSeqs)</pre>
<div class="block">Removes pairs of reads where at least one of the pair contains short/long
 reads or a read that contain an 'N', from an interlaced fastq file</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fastqFileIn</code> - an interlaced fastq file</dd><dd><code>leftReadsOut</code> - the good left-handed reads</dd><dd><code>rightReadsOut</code> - the good right-handed reads</dd><dd><code>singleEndReadLength</code> - the expected length of a single read</dd><dd><code>format</code> - the fastq format (can only be 'illumina' or 'sanger')</dd><dd><code>writeBadSeqs</code> - whether to write the bad reads to a file (bad reads
 file name will start with 'bad_')</dd></dl>
</li>
</ul>
<a name="qcJoinedReads(java.io.File, java.io.File, java.io.File, int, java.lang.String, boolean)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>qcJoinedReads</h4>
<pre>public&nbsp;void&nbsp;qcJoinedReads(java.io.File&nbsp;fastqFileIn,
                 java.io.File&nbsp;leftReadsOut,
                 java.io.File&nbsp;rightReadsOut,
                 int&nbsp;readLength,
                 java.lang.String&nbsp;format,
                 boolean&nbsp;writeBadSeqs)</pre>
<div class="block">Removes pairs of reads where at least one of the pair contains short/long
 reads or a read that contain an 'N', from a fastq file where the reads
 are joined</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fastqFileIn</code> - an interlaced fastq file</dd><dd><code>leftReadsOut</code> - the good left-handed reads</dd><dd><code>rightReadsOut</code> - the good right-handed reads</dd><dd><code>readLength</code> - the expected length of a single read (i.e. half the
 length of one of the joined reads)</dd><dd><code>format</code> - the fastq format (can only be 'illumina' or 'sanger')</dd><dd><code>writeBadSeqs</code> - whether to write the bad reads to a file (bad reads
 file name will start with 'bad_')</dd></dl>
</li>
</ul>
<a name="qcSingleEndReads(java.io.File, java.io.File, int, java.lang.String, boolean)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>qcSingleEndReads</h4>
<pre>public&nbsp;void&nbsp;qcSingleEndReads(java.io.File&nbsp;fastqFileIn,
                    java.io.File&nbsp;readsOut,
                    int&nbsp;singleEndReadLength,
                    java.lang.String&nbsp;format,
                    boolean&nbsp;writeBadSeqs)</pre>
<div class="block">Removes reads that contain short/long reads or that contain an 'N', from
 a fastq file</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fastqFileIn</code> - the fastq file</dd><dd><code>readsOut</code> - the good reads</dd><dd><code>singleEndReadLength</code> - the expected length of a single read</dd><dd><code>format</code> - the fastq format (can only be 'illumina' or 'sanger')</dd><dd><code>writeBadSeqs</code> - whether to write the bad reads to a file (bad reads
 file name will start with 'bad_')</dd></dl>
</li>
</ul>
<a name="groomRead(net.sf.picard.fastq.FastqRecord, java.lang.String)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>groomRead</h4>
<pre>public&nbsp;net.sf.picard.fastq.FastqRecord&nbsp;groomRead(net.sf.picard.fastq.FastqRecord&nbsp;record,
                                        java.lang.String&nbsp;format)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>record</code> - the FastqRecord to be groomed</dd><dd><code>format</code> - the current format. Can only be 'illumina' or 'sanger'</dd>
<dt><span class="strong">Returns:</span></dt><dd>a FastqRecord in sanger format</dd></dl>
</li>
</ul>
<a name="guessFormat(net.sf.picard.fastq.FastqRecord)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>guessFormat</h4>
<pre>public&nbsp;net.sf.picard.util.FastqQualityFormat&nbsp;guessFormat(net.sf.picard.fastq.FastqRecord&nbsp;record)</pre>
<div class="block">Guesses the format of a FastqRecord</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>record</code> - a FastqRecord of unknown format</dd>
<dt><span class="strong">Returns:</span></dt><dd>the best guess;</dd></dl>
</li>
</ul>
<a name="countLengthsAndNs(java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>countLengthsAndNs</h4>
<pre>public&nbsp;int&nbsp;countLengthsAndNs(java.io.File&nbsp;fastq)</pre>
<div class="block">Takes a fastq file and calculates the number of reads which have a 'N'
 and the distribution of read lengths</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fastq</code> - the fastq file to analyse</dd></dl>
</li>
</ul>
<a name="veryfiyReads(java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>veryfiyReads</h4>
<pre>public&nbsp;void&nbsp;veryfiyReads(java.io.File&nbsp;fastq)</pre>
<div class="block">Checks a fastq file to verify that all the reads can be parsed into a
 FastqRecord. Errors should be thrown if any of the reads are not
 formatted correctly.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fastq</code> - the fastq file to verify</dd></dl>
</li>
</ul>
<a name="veryfiyPairedReads(java.io.File, java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>veryfiyPairedReads</h4>
<pre>public&nbsp;void&nbsp;veryfiyPairedReads(java.io.File&nbsp;fastqLeft,
                      java.io.File&nbsp;fastqRight)</pre>
<div class="block">Verifies that two files of left/right paired-end reads are in order and
 contain no fastq format errors</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fastqLeft</code> - the left-handed fastq file to verify</dd><dd><code>fastqRight</code> - the right-handed fastq file to verify</dd></dl>
</li>
</ul>
<a name="main(java.lang.String[])">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>main</h4>
<pre>public static&nbsp;void&nbsp;main(java.lang.String[]&nbsp;args)</pre>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar_bottom">
<!--   -->
</a><a href="#skip-navbar_bottom" title="Skip navigation links"></a><a name="navbar_bottom_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/FastqQC.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html" title="class in uk.ac.tsl.etherington.genomehelper.fastq"><span class="strong">Prev Class</span></a></li>
<li>Next Class</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html" target="_top">Frames</a></li>
<li><a href="FastqQC.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
