<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (version 1.7.0_21) on Tue Oct 29 14:06:34 GMT 2013 -->
<meta http-equiv="Content-Type" content="text/html" charset="UTF-8">
<title>FastqParser (GenomeHelper)</title>
<meta name="date" content="2013-10-29">
<link rel="stylesheet" type="text/css" href="../../../../../../stylesheet.css" title="Style">
</head>
<body>
<script type="text/javascript"><!--
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="FastqParser (GenomeHelper)";
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar_top">
<!--   -->
</a><a href="#skip-navbar_top" title="Skip navigation links"></a><a name="navbar_top_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/FastqParser.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqMotifFinder.StringLengthComparator.html" title="class in uk.ac.tsl.etherington.genomehelper.fastq"><span class="strong">Prev Class</span></a></li>
<li><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html" title="class in uk.ac.tsl.etherington.genomehelper.fastq"><span class="strong">Next Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html" target="_top">Frames</a></li>
<li><a href="FastqParser.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">uk.ac.tsl.etherington.genomehelper.fastq</div>
<h2 title="Class FastqParser" class="title">Class FastqParser</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>uk.ac.tsl.etherington.genomehelper.fastq.FastqParser</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<hr>
<br>
<pre>public class <span class="strong">FastqParser</span>
extends java.lang.Object</pre>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html#FastqParser()">FastqParser</a></strong>()</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method_summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span>Methods</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html#fastqToFastaFile(java.io.File, java.io.File)">fastqToFastaFile</a></strong>(java.io.File&nbsp;fastqIn,
                java.io.File&nbsp;fastaOut)</code>
<div class="block">Turns a fastq file into fasta format</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>org.biojava3.data.sequence.FastaSequence</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html#fastqToFastaSeq(net.sf.picard.fastq.FastqRecord)">fastqToFastaSeq</a></strong>(net.sf.picard.fastq.FastqRecord&nbsp;record)</code>
<div class="block">Takes a net.sf.picard.fastq.FastqRecord and changes it into a
 org.biojava3.data.sequence.FastqSequence</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html#fastqToFastaSixFrameTranslation(java.io.File, java.io.File, boolean)">fastqToFastaSixFrameTranslation</a></strong>(java.io.File&nbsp;fastqIn,
                               java.io.File&nbsp;fastaOut,
                               boolean&nbsp;includeDNASeq)</code>
<div class="block">Creates a file of six-frame translations (in fasta format) from a fastq
 file</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html#getOneSideFastqSeqsFromList(java.io.File, java.io.File, java.io.File)">getOneSideFastqSeqsFromList</a></strong>(java.io.File&nbsp;listFile,
                           java.io.File&nbsp;fastqFileIn,
                           java.io.File&nbsp;fastqFileOut)</code>
<div class="block">Extract fastq sequences in a list from a file of single-end fastq
 sequences</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html#getPairedFastqSeqsFromHashSet(java.util.HashSet, java.io.File, java.io.File, java.io.File, java.io.File)">getPairedFastqSeqsFromHashSet</a></strong>(java.util.HashSet&nbsp;list,
                             java.io.File&nbsp;fastqLeft,
                             java.io.File&nbsp;fastqRight,
                             java.io.File&nbsp;fastqLeftOut,
                             java.io.File&nbsp;fastqRightOut)</code>
<div class="block">Extract paired-end fastq sequences from a list of sequence names</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.util.HashSet</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html#readNamesToHashSet(java.io.File)">readNamesToHashSet</a></strong>(java.io.File&nbsp;readNames)</code>
<div class="block">Returns a HashSet of fastq read names.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html#writeRecords(net.sf.picard.fastq.FastqReader, net.sf.picard.fastq.FastqWriter, java.util.HashSet)">writeRecords</a></strong>(net.sf.picard.fastq.FastqReader&nbsp;fastqReader,
            net.sf.picard.fastq.FastqWriter&nbsp;out,
            java.util.HashSet&lt;java.lang.String&gt;&nbsp;mappedReads)</code>&nbsp;</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods_inherited_from_class_java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="FastqParser()">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>FastqParser</h4>
<pre>public&nbsp;FastqParser()</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method_detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="readNamesToHashSet(java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>readNamesToHashSet</h4>
<pre>public&nbsp;java.util.HashSet&nbsp;readNamesToHashSet(java.io.File&nbsp;readNames)
                                     throws java.io.FileNotFoundException,
                                            java.io.IOException</pre>
<div class="block">Returns a HashSet of fastq read names.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>readNames</code> - a text file of read names, one per line</dd>
<dt><span class="strong">Returns:</span></dt><dd>HashSet of read names</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.FileNotFoundException</code></dd>
<dd><code>java.io.IOException</code></dd></dl>
</li>
</ul>
<a name="writeRecords(net.sf.picard.fastq.FastqReader, net.sf.picard.fastq.FastqWriter, java.util.HashSet)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>writeRecords</h4>
<pre>public static&nbsp;void&nbsp;writeRecords(net.sf.picard.fastq.FastqReader&nbsp;fastqReader,
                net.sf.picard.fastq.FastqWriter&nbsp;out,
                java.util.HashSet&lt;java.lang.String&gt;&nbsp;mappedReads)</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fastqReader</code> - </dd><dd><code>out</code> - </dd><dd><code>mappedReads</code> - </dd></dl>
</li>
</ul>
<a name="getPairedFastqSeqsFromHashSet(java.util.HashSet, java.io.File, java.io.File, java.io.File, java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getPairedFastqSeqsFromHashSet</h4>
<pre>public&nbsp;void&nbsp;getPairedFastqSeqsFromHashSet(java.util.HashSet&nbsp;list,
                                 java.io.File&nbsp;fastqLeft,
                                 java.io.File&nbsp;fastqRight,
                                 java.io.File&nbsp;fastqLeftOut,
                                 java.io.File&nbsp;fastqRightOut)
                                   throws java.io.FileNotFoundException,
                                          java.io.IOException</pre>
<div class="block">Extract paired-end fastq sequences from a list of sequence names</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>list</code> - a HashSet of read names</dd><dd><code>fastqFileInLeft</code> - all left-handed reads</dd><dd><code>fastqFileInRight</code> - all right-handed reads</dd><dd><code>fastqFileOutLeft</code> - the left-handed reads in the list</dd><dd><code>fastqFileOutRight</code> - the right-handed reads in the list</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.FileNotFoundException</code></dd>
<dd><code>java.io.IOException</code></dd></dl>
</li>
</ul>
<a name="getOneSideFastqSeqsFromList(java.io.File, java.io.File, java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getOneSideFastqSeqsFromList</h4>
<pre>public&nbsp;void&nbsp;getOneSideFastqSeqsFromList(java.io.File&nbsp;listFile,
                               java.io.File&nbsp;fastqFileIn,
                               java.io.File&nbsp;fastqFileOut)
                                 throws java.io.FileNotFoundException,
                                        java.io.IOException</pre>
<div class="block">Extract fastq sequences in a list from a file of single-end fastq
 sequences</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>listFile</code> - a file of read names, one per line</dd><dd><code>fastqFileIn</code> - the fastq file to search for the read names</dd><dd><code>fastqFileOut</code> - the fastq file of found read names</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.FileNotFoundException</code></dd>
<dd><code>java.io.IOException</code></dd></dl>
</li>
</ul>
<a name="fastqToFastaFile(java.io.File, java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>fastqToFastaFile</h4>
<pre>public&nbsp;void&nbsp;fastqToFastaFile(java.io.File&nbsp;fastqIn,
                    java.io.File&nbsp;fastaOut)
                      throws java.io.IOException</pre>
<div class="block">Turns a fastq file into fasta format</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fastqIn</code> - the fastq intput file</dd><dd><code>fastaOut</code> - the fasta output file</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd></dl>
</li>
</ul>
<a name="fastqToFastaSeq(net.sf.picard.fastq.FastqRecord)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>fastqToFastaSeq</h4>
<pre>public&nbsp;org.biojava3.data.sequence.FastaSequence&nbsp;fastqToFastaSeq(net.sf.picard.fastq.FastqRecord&nbsp;record)
                                                         throws java.io.IOException</pre>
<div class="block">Takes a net.sf.picard.fastq.FastqRecord and changes it into a
 org.biojava3.data.sequence.FastqSequence</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>record</code> - the FastqRecord to change</dd>
<dt><span class="strong">Returns:</span></dt><dd>a FastaSequence object</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd></dl>
</li>
</ul>
<a name="fastqToFastaSixFrameTranslation(java.io.File, java.io.File, boolean)">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>fastqToFastaSixFrameTranslation</h4>
<pre>public&nbsp;void&nbsp;fastqToFastaSixFrameTranslation(java.io.File&nbsp;fastqIn,
                                   java.io.File&nbsp;fastaOut,
                                   boolean&nbsp;includeDNASeq)
                                     throws java.io.IOException</pre>
<div class="block">Creates a file of six-frame translations (in fasta format) from a fastq
 file</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fastqIn</code> - the fastq file to translate</dd><dd><code>fastaOut</code> - the translation file</dd><dd><code>includeDNASeq</code> - set to true if the original DNA sequence is required
 in the outfile</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd></dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar_bottom">
<!--   -->
</a><a href="#skip-navbar_bottom" title="Skip navigation links"></a><a name="navbar_bottom_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/FastqParser.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqMotifFinder.StringLengthComparator.html" title="class in uk.ac.tsl.etherington.genomehelper.fastq"><span class="strong">Prev Class</span></a></li>
<li><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/fastq/FastqQC.html" title="class in uk.ac.tsl.etherington.genomehelper.fastq"><span class="strong">Next Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?uk/ac/tsl/etherington/genomehelper/fastq/FastqParser.html" target="_top">Frames</a></li>
<li><a href="FastqParser.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
