<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (version 1.7.0_21) on Wed Dec 18 15:53:34 GMT 2013 -->
<meta http-equiv="Content-Type" content="text/html" charset="UTF-8">
<title>MappedSamRecords (GenomeHelper)</title>
<meta name="date" content="2013-12-18">
<link rel="stylesheet" type="text/css" href="../../../../../../stylesheet.css" title="Style">
</head>
<body>
<script type="text/javascript"><!--
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="MappedSamRecords (GenomeHelper)";
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar_top">
<!--   -->
</a><a href="#skip-navbar_top" title="Skip navigation links"></a><a name="navbar_top_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/MappedSamRecords.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev Class</li>
<li>Next Class</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html" target="_top">Frames</a></li>
<li><a href="MappedSamRecords.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">uk.ac.tsl.etherington.genomehelper.bam</div>
<h2 title="Class MappedSamRecords" class="title">Class MappedSamRecords</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>uk.ac.tsl.etherington.genomehelper.bam.MappedSamRecords</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<hr>
<br>
<pre>public class <span class="strong">MappedSamRecords</span>
extends java.lang.Object</pre>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#MappedSamRecords()">MappedSamRecords</a></strong>()</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method_summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span>Methods</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getBothPairedMappedSamRecords(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File)">getBothPairedMappedSamRecords</a></strong>(java.io.File&nbsp;bamFile,
                             java.io.File&nbsp;fastqInLeft,
                             java.io.File&nbsp;fastqInRight,
                             java.io.File&nbsp;fastqOutLeft,
                             java.io.File&nbsp;fastqOutRight)</code>
<div class="block">Returns the pairs of sequences where at both of the pair has been mapped
 to a genome.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getBothPairedUnmappedSamRecords(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File)">getBothPairedUnmappedSamRecords</a></strong>(java.io.File&nbsp;bamFile,
                               java.io.File&nbsp;fastqInLeft,
                               java.io.File&nbsp;fastqInRight,
                               java.io.File&nbsp;fastqOutLeft,
                               java.io.File&nbsp;fastqOutRight)</code>
<div class="block">Returns the pairs of sequences where both of the pair has not been mapped
 to a genome.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getOnePairedMappedSamRecords(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File)">getOnePairedMappedSamRecords</a></strong>(java.io.File&nbsp;bamFile,
                            java.io.File&nbsp;fastqInLeft,
                            java.io.File&nbsp;fastqInRight,
                            java.io.File&nbsp;fastqOutLeft,
                            java.io.File&nbsp;fastqOutRight)</code>
<div class="block">Returns the pairs of sequences where at least one of the pair has been
 mapped to a genome.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getOnePairedUnmappedSamRecords(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File)">getOnePairedUnmappedSamRecords</a></strong>(java.io.File&nbsp;bamFile,
                              java.io.File&nbsp;fastqInLeft,
                              java.io.File&nbsp;fastqInRight,
                              java.io.File&nbsp;fastqOutLeft,
                              java.io.File&nbsp;fastqOutRight)</code>
<div class="block">Returns the pairs of sequences where at least one of the pair has not
 been mapped to a genome.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.util.HashSet&lt;java.lang.String&gt;</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getReads(net.sf.samtools.SAMFileReader, net.sf.samtools.SAMRecordIterator, boolean)">getReads</a></strong>(net.sf.samtools.SAMFileReader&nbsp;samReader,
        net.sf.samtools.SAMRecordIterator&nbsp;iterator,
        boolean&nbsp;getMapped)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getSingleMappedSamRecords(java.io.File, java.io.File, java.io.File, boolean)">getSingleMappedSamRecords</a></strong>(java.io.File&nbsp;bamFile,
                         java.io.File&nbsp;fastqIn,
                         java.io.File&nbsp;fastqOut,
                         boolean&nbsp;isRightHandedReads)</code>
<div class="block">Returns the fastq sequences that have been mapped to a genome.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html#getSingleUnmappedSamRecords(java.io.File, java.io.File, java.io.File, boolean)">getSingleUnmappedSamRecords</a></strong>(java.io.File&nbsp;bamFile,
                           java.io.File&nbsp;fastqIn,
                           java.io.File&nbsp;fastqOut,
                           boolean&nbsp;isRightHandedReads)</code>
<div class="block">Returns the fastq sequences that have not been mapped to a genome.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods_inherited_from_class_java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="MappedSamRecords()">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>MappedSamRecords</h4>
<pre>public&nbsp;MappedSamRecords()</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method_detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="getSingleMappedSamRecords(java.io.File, java.io.File, java.io.File, boolean)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSingleMappedSamRecords</h4>
<pre>public&nbsp;void&nbsp;getSingleMappedSamRecords(java.io.File&nbsp;bamFile,
                             java.io.File&nbsp;fastqIn,
                             java.io.File&nbsp;fastqOut,
                             boolean&nbsp;isRightHandedReads)</pre>
<div class="block">Returns the fastq sequences that have been mapped to a genome. All fastq
 reads must either be left-handed or right-handed</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>bamFile</code> - a sam/bam file</dd><dd><code>fastqIn</code> - the fastq dataset that was mapped to the sam/bam file</dd><dd><code>fastqOut</code> - the fastq reads that were actually mapped</dd><dd><code>isRightHandedReads</code> - whether the reads are right-handed</dd></dl>
</li>
</ul>
<a name="getSingleUnmappedSamRecords(java.io.File, java.io.File, java.io.File, boolean)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSingleUnmappedSamRecords</h4>
<pre>public&nbsp;void&nbsp;getSingleUnmappedSamRecords(java.io.File&nbsp;bamFile,
                               java.io.File&nbsp;fastqIn,
                               java.io.File&nbsp;fastqOut,
                               boolean&nbsp;isRightHandedReads)</pre>
<div class="block">Returns the fastq sequences that have not been mapped to a genome. All
 fastq reads must either be left-handed or right-handed</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>bamFile</code> - a sam/bam file</dd><dd><code>fastqIn</code> - the fastq dataset that was mapped to the sam/bam file</dd><dd><code>fastqOut</code> - the fastq reads that were not mapped</dd><dd><code>isRightHandedReads</code> - whether the reads are right-handed</dd></dl>
</li>
</ul>
<a name="getOnePairedMappedSamRecords(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getOnePairedMappedSamRecords</h4>
<pre>public&nbsp;void&nbsp;getOnePairedMappedSamRecords(java.io.File&nbsp;bamFile,
                                java.io.File&nbsp;fastqInLeft,
                                java.io.File&nbsp;fastqInRight,
                                java.io.File&nbsp;fastqOutLeft,
                                java.io.File&nbsp;fastqOutRight)</pre>
<div class="block">Returns the pairs of sequences where at least one of the pair has been
 mapped to a genome.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>bamFile</code> - a sam/bam file</dd><dd><code>fastqInLeft</code> - the left-hand fastq dataset that was mapped to the
 sam/bam file</dd><dd><code>fastqInRight</code> - the righ-hand fastq dataset that was mapped to the
 sam/bam file</dd><dd><code>fastqOutLeft</code> - the left-hand fastq reads where one of the pair
 actually mapped</dd><dd><code>fastqOutRight</code> - the right-hand fastq reads where one of the pair
 actually mapped</dd></dl>
</li>
</ul>
<a name="getReads(net.sf.samtools.SAMFileReader, net.sf.samtools.SAMRecordIterator, boolean)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getReads</h4>
<pre>public&nbsp;java.util.HashSet&lt;java.lang.String&gt;&nbsp;getReads(net.sf.samtools.SAMFileReader&nbsp;samReader,
                                           net.sf.samtools.SAMRecordIterator&nbsp;iterator,
                                           boolean&nbsp;getMapped)</pre>
</li>
</ul>
<a name="getOnePairedUnmappedSamRecords(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getOnePairedUnmappedSamRecords</h4>
<pre>public&nbsp;void&nbsp;getOnePairedUnmappedSamRecords(java.io.File&nbsp;bamFile,
                                  java.io.File&nbsp;fastqInLeft,
                                  java.io.File&nbsp;fastqInRight,
                                  java.io.File&nbsp;fastqOutLeft,
                                  java.io.File&nbsp;fastqOutRight)</pre>
<div class="block">Returns the pairs of sequences where at least one of the pair has not
 been mapped to a genome.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>bamFile</code> - a sam/bam file</dd><dd><code>fastqInLeft</code> - the left-hand fastq dataset that was mapped to the
 sam/bam file</dd><dd><code>fastqInRight</code> - the righ-hand fastq dataset that was mapped to the
 sam/bam file</dd><dd><code>fastqOutLeft</code> - the left-hand fastq reads where one of the pair was
 not mapped</dd><dd><code>fastqOutRight</code> - the right-hand fastq reads where one of the pair was
 not mapped</dd></dl>
</li>
</ul>
<a name="getBothPairedMappedSamRecords(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getBothPairedMappedSamRecords</h4>
<pre>public&nbsp;void&nbsp;getBothPairedMappedSamRecords(java.io.File&nbsp;bamFile,
                                 java.io.File&nbsp;fastqInLeft,
                                 java.io.File&nbsp;fastqInRight,
                                 java.io.File&nbsp;fastqOutLeft,
                                 java.io.File&nbsp;fastqOutRight)</pre>
<div class="block">Returns the pairs of sequences where at both of the pair has been mapped
 to a genome.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>bamFile</code> - a sam/bam file</dd><dd><code>fastqInLeft</code> - the left-hand fastq dataset that was mapped to the
 sam/bam file</dd><dd><code>fastqInRight</code> - the righ-hand fastq dataset that was mapped to the
 sam/bam file</dd><dd><code>fastqOutLeft</code> - the left-hand fastq reads where both of the pairs
 actually mapped</dd><dd><code>fastqOutRight</code> - the right-hand fastq reads where both of the pairs
 actually mapped</dd></dl>
</li>
</ul>
<a name="getBothPairedUnmappedSamRecords(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File)">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>getBothPairedUnmappedSamRecords</h4>
<pre>public&nbsp;void&nbsp;getBothPairedUnmappedSamRecords(java.io.File&nbsp;bamFile,
                                   java.io.File&nbsp;fastqInLeft,
                                   java.io.File&nbsp;fastqInRight,
                                   java.io.File&nbsp;fastqOutLeft,
                                   java.io.File&nbsp;fastqOutRight)</pre>
<div class="block">Returns the pairs of sequences where both of the pair has not been mapped
 to a genome.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>bamFile</code> - a sam/bam file</dd><dd><code>fastqInLeft</code> - the left-hand fastq dataset that was mapped to the
 sam/bam file</dd><dd><code>fastqInRight</code> - the righ-hand fastq dataset that was mapped to the
 sam/bam file</dd><dd><code>fastqOutLeft</code> - the left-hand fastq reads where both of the pairs was
 not mapped</dd><dd><code>fastqOutRight</code> - the right-hand fastq reads where both of the pairs
 was not mapped</dd></dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar_bottom">
<!--   -->
</a><a href="#skip-navbar_bottom" title="Skip navigation links"></a><a name="navbar_bottom_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/MappedSamRecords.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev Class</li>
<li>Next Class</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?uk/ac/tsl/etherington/genomehelper/bam/MappedSamRecords.html" target="_top">Frames</a></li>
<li><a href="MappedSamRecords.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
