<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (version 1.7.0_21) on Wed Dec 18 15:53:33 GMT 2013 -->
<meta http-equiv="Content-Type" content="text/html" charset="UTF-8">
<title>GFFFeatureStats (GenomeHelper)</title>
<meta name="date" content="2013-12-18">
<link rel="stylesheet" type="text/css" href="../../../../../../stylesheet.css" title="Style">
</head>
<body>
<script type="text/javascript"><!--
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="GFFFeatureStats (GenomeHelper)";
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar_top">
<!--   -->
</a><a href="#skip-navbar_top" title="Skip navigation links"></a><a name="navbar_top_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/GFFFeatureStats.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev Class</li>
<li>Next Class</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" target="_top">Frames</a></li>
<li><a href="GFFFeatureStats.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">uk.ac.tsl.etherington.genomehelper.gff</div>
<h2 title="Class GFFFeatureStats" class="title">Class GFFFeatureStats</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>uk.ac.tsl.etherington.genomehelper.gff.GFFFeatureStats</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<hr>
<br>
<pre>public class <span class="strong">GFFFeatureStats</span>
extends java.lang.Object</pre>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#GFFFeatureStats()">GFFFeatureStats</a></strong>()</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method_summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span>Methods</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#calculateCodingRegion(org.biojava3.genome.parsers.gff.FeatureList, java.util.HashMap, java.lang.String, double)">calculateCodingRegion</a></strong>(org.biojava3.genome.parsers.gff.FeatureList&nbsp;featList,
                     java.util.HashMap&lt;java.lang.String,int[]&gt;&nbsp;codingMap,
                     java.lang.String&nbsp;attribute,
                     double&nbsp;genomeLength)</code>
<div class="block">Calculate the proportion of the genome that contains coding regions (from
 the start of the first exon to the end of the last exon)</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#createCodingGenome(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, java.io.File, java.io.File)">createCodingGenome</a></strong>(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                  java.lang.String&nbsp;feature,
                  java.io.File&nbsp;refSeq,
                  java.io.File&nbsp;codingGenome)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#createNonCodingGenome(org.biojava3.genome.parsers.gff.FeatureList, java.io.File, java.io.File)">createNonCodingGenome</a></strong>(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                     java.io.File&nbsp;refSeq,
                     java.io.File&nbsp;nonCodingGenome)</code>
<div class="block">Writes the DNA sequence of the non-coding portion of the genome (i.e.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.util.HashMap&lt;java.lang.String,java.util.ArrayList&lt;java.lang.Integer&gt;&gt;</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getBlocks(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String)">getBlocks</a></strong>(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
         java.lang.String&nbsp;featureType)</code>
<div class="block">Creates start/end blocks of features</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.util.HashMap&lt;java.lang.String,java.util.ArrayList&lt;java.lang.Integer&gt;&gt;</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getBlocks(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, java.lang.String)">getBlocks</a></strong>(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
         java.lang.String&nbsp;featureType,
         java.lang.String&nbsp;attribute)</code>
<div class="block">Creates start/end blocks of features</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.util.HashMap&lt;java.lang.String,java.util.ArrayList&lt;java.lang.Integer&gt;&gt;</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getBlocks(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, java.lang.String, java.util.Set)">getBlocks</a></strong>(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
         java.lang.String&nbsp;featureType,
         java.lang.String&nbsp;attribute,
         java.util.Set&nbsp;secretedProteins)</code>
<div class="block">Creates start/end blocks of features</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>org.biojava3.genome.parsers.gff.FeatureList</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getFeatureList(java.lang.String)">getFeatureList</a></strong>(java.lang.String&nbsp;gffFile)</code>
<div class="block">Gets all the features from a gff file</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getGenomeSizeFromIntArrayHashMap(java.util.HashMap)">getGenomeSizeFromIntArrayHashMap</a></strong>(java.util.HashMap&lt;java.lang.String,int[]&gt;&nbsp;genomeMap)</code>
<div class="block">Calculates the length of a genome from collection of int arrays, where
 each int array is the length of a genome sequence</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getMeanFeatureLength(org.biojava3.genome.parsers.gff.FeatureList, java.util.HashMap, java.lang.String)">getMeanFeatureLength</a></strong>(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                    java.util.HashMap&lt;java.lang.String,int[]&gt;&nbsp;genomeMap,
                    java.lang.String&nbsp;featureName)</code>
<div class="block">Prints stats and mean feature length for a given feature</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getMeanFeatureLength(org.biojava3.genome.parsers.gff.FeatureList, java.util.HashMap, java.lang.String, java.io.File, java.lang.String)">getMeanFeatureLength</a></strong>(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                    java.util.HashMap&lt;java.lang.String,int[]&gt;&nbsp;genomeMap,
                    java.lang.String&nbsp;featureName,
                    java.io.File&nbsp;geneIds,
                    java.lang.String&nbsp;attribute)</code>
<div class="block">Prints stats and mean feature length for a given feature, but restricts
 it to a list of geneIds of interest</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getMeanFeatureLength(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String)">getMeanFeatureLength</a></strong>(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                    java.lang.String&nbsp;featureName)</code>
<div class="block">Calculates the mean feature length of a given feature</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getMeanIntronLength(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, double)">getMeanIntronLength</a></strong>(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                   java.lang.String&nbsp;attribute,
                   double&nbsp;genomeSize)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getMeanSecretedIntronLength(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, java.io.File)">getMeanSecretedIntronLength</a></strong>(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                           java.lang.String&nbsp;attribute,
                           java.io.File&nbsp;secretedProteinsFile)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../../../../../uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html#getStats(java.lang.String, java.io.File, java.lang.String)">getStats</a></strong>(java.lang.String&nbsp;gff,
        java.io.File&nbsp;refSeq,
        java.lang.String&nbsp;attribute)</code>
<div class="block">Combines a number of methods</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods_inherited_from_class_java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="GFFFeatureStats()">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>GFFFeatureStats</h4>
<pre>public&nbsp;GFFFeatureStats()</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method_detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="getStats(java.lang.String, java.io.File, java.lang.String)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getStats</h4>
<pre>public&nbsp;void&nbsp;getStats(java.lang.String&nbsp;gff,
            java.io.File&nbsp;refSeq,
            java.lang.String&nbsp;attribute)
              throws java.io.FileNotFoundException,
                     org.biojava.bio.BioException,
                     java.io.IOException,
                     java.lang.Exception</pre>
<div class="block">Combines a number of methods</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>gff</code> - </dd><dd><code>refSeq</code> - </dd><dd><code>attribute</code> - </dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.FileNotFoundException</code></dd>
<dd><code>org.biojava.bio.BioException</code></dd>
<dd><code>java.io.IOException</code></dd>
<dd><code>java.lang.Exception</code></dd></dl>
</li>
</ul>
<a name="getGenomeSizeFromIntArrayHashMap(java.util.HashMap)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getGenomeSizeFromIntArrayHashMap</h4>
<pre>public&nbsp;double&nbsp;getGenomeSizeFromIntArrayHashMap(java.util.HashMap&lt;java.lang.String,int[]&gt;&nbsp;genomeMap)</pre>
<div class="block">Calculates the length of a genome from collection of int arrays, where
 each int array is the length of a genome sequence</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>genomeMap</code> - </dd>
<dt><span class="strong">Returns:</span></dt><dd>the size of the genome (in nucleotides)</dd></dl>
</li>
</ul>
<a name="getFeatureList(java.lang.String)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getFeatureList</h4>
<pre>public&nbsp;org.biojava3.genome.parsers.gff.FeatureList&nbsp;getFeatureList(java.lang.String&nbsp;gffFile)
                                                           throws java.io.IOException</pre>
<div class="block">Gets all the features from a gff file</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>gffFile</code> - the gff file from which to extract the features</dd>
<dt><span class="strong">Returns:</span></dt><dd>a org.biojava3.genome.parsers.gff.FeatureList</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd></dl>
</li>
</ul>
<a name="getMeanFeatureLength(org.biojava3.genome.parsers.gff.FeatureList, java.util.HashMap, java.lang.String)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMeanFeatureLength</h4>
<pre>public&nbsp;double&nbsp;getMeanFeatureLength(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                          java.util.HashMap&lt;java.lang.String,int[]&gt;&nbsp;genomeMap,
                          java.lang.String&nbsp;featureName)
                            throws java.io.IOException,
                                   java.lang.Exception</pre>
<div class="block">Prints stats and mean feature length for a given feature</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fl</code> - a org.biojava3.genome.parsers.gff.FeatureList</dd><dd><code>genomeMap</code> - a HashMap where the keys are the reference sequence
 names and the values a zero-filled int [] the length of the associated
 reference sequence</dd><dd><code>featureName</code> - the feature-type to analyse (e.g. mRNA, exon, CDS,
 etc)</dd>
<dt><span class="strong">Returns:</span></dt><dd>the mean length of the requested feature name</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd>
<dd><code>java.lang.Exception</code></dd></dl>
</li>
</ul>
<a name="getMeanFeatureLength(org.biojava3.genome.parsers.gff.FeatureList, java.util.HashMap, java.lang.String, java.io.File, java.lang.String)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMeanFeatureLength</h4>
<pre>public&nbsp;double&nbsp;getMeanFeatureLength(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                          java.util.HashMap&lt;java.lang.String,int[]&gt;&nbsp;genomeMap,
                          java.lang.String&nbsp;featureName,
                          java.io.File&nbsp;geneIds,
                          java.lang.String&nbsp;attribute)
                            throws java.io.IOException,
                                   java.lang.Exception</pre>
<div class="block">Prints stats and mean feature length for a given feature, but restricts
 it to a list of geneIds of interest</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fl</code> - a org.biojava3.genome.parsers.gff.FeatureList</dd><dd><code>genomeMap</code> - a HashMap where the keys are the reference sequence
 names and the values a zero-filled int [] the length of the associated
 reference sequence</dd><dd><code>featureName</code> - the feature-type to analyse (e.g. mRNA, exon, CDS,
 etc)</dd><dd><code>geneIds</code> - a file of gene IDs, one per line</dd><dd><code>attribute</code> - the gff attribute that the gene IDs will be listed
 under, e.g. gene_id, transcript_id, etc</dd>
<dt><span class="strong">Returns:</span></dt><dd>prints out stats and returns the mean feature length of the
 featureName</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd>
<dd><code>java.lang.Exception</code></dd></dl>
</li>
</ul>
<a name="getMeanFeatureLength(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMeanFeatureLength</h4>
<pre>public&nbsp;double&nbsp;getMeanFeatureLength(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                          java.lang.String&nbsp;featureName)
                            throws java.io.IOException</pre>
<div class="block">Calculates the mean feature length of a given feature</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fl</code> - a org.biojava3.genome.parsers.gff.FeatureList</dd><dd><code>featureName</code> - the feature-type to analyse (e.g. mRNA, exon, CDS,
 etc)</dd>
<dt><span class="strong">Returns:</span></dt><dd>the mean feature length of the given feature</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd></dl>
</li>
</ul>
<a name="createNonCodingGenome(org.biojava3.genome.parsers.gff.FeatureList, java.io.File, java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createNonCodingGenome</h4>
<pre>public&nbsp;void&nbsp;createNonCodingGenome(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                         java.io.File&nbsp;refSeq,
                         java.io.File&nbsp;nonCodingGenome)
                           throws java.lang.Exception</pre>
<div class="block">Writes the DNA sequence of the non-coding portion of the genome (i.e.
 intergenic and intron).</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fl</code> - a org.biojava3.genome.parsers.gff.FeatureList</dd><dd><code>refSeq</code> - the reference sequence (in fasta format)</dd><dd><code>nonCodingGenome</code> - the non-coding reference geneom (in fasta format)</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.lang.Exception</code></dd></dl>
</li>
</ul>
<a name="createCodingGenome(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, java.io.File, java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createCodingGenome</h4>
<pre>public&nbsp;void&nbsp;createCodingGenome(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                      java.lang.String&nbsp;feature,
                      java.io.File&nbsp;refSeq,
                      java.io.File&nbsp;codingGenome)
                        throws java.lang.Exception</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fl</code> - a org.biojava3.genome.parsers.gff.FeatureList</dd><dd><code>feature</code> - the feature to use for the coding genome (e.g. exon, cds,
 mRNA)</dd><dd><code>refSeq</code> - the reference sequence (in fasta format)</dd><dd><code>codingGenome</code> - the non-coding reference genome (in fasta format)</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.lang.Exception</code></dd></dl>
</li>
</ul>
<a name="getMeanIntronLength(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, double)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMeanIntronLength</h4>
<pre>public&nbsp;double&nbsp;getMeanIntronLength(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                         java.lang.String&nbsp;attribute,
                         double&nbsp;genomeSize)
                           throws java.io.IOException,
                                  java.io.FileNotFoundException,
                                  org.biojava.bio.BioException</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fl</code> - a org.biojava3.genome.parsers.gff.FeatureList</dd><dd><code>attribute</code> - the name of the attribute that will make your genes
 unique (e.g. "name", gene_id, etc)</dd><dd><code>genomeSize</code> - the size of the genome</dd>
<dt><span class="strong">Returns:</span></dt><dd>the mean intron length</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd>
<dd><code>java.io.FileNotFoundException</code></dd>
<dd><code>org.biojava.bio.BioException</code></dd></dl>
</li>
</ul>
<a name="getMeanSecretedIntronLength(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, java.io.File)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMeanSecretedIntronLength</h4>
<pre>public&nbsp;double&nbsp;getMeanSecretedIntronLength(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                                 java.lang.String&nbsp;attribute,
                                 java.io.File&nbsp;secretedProteinsFile)
                                   throws java.io.IOException</pre>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fl</code> - a org.biojava3.genome.parsers.gff.FeatureList</dd><dd><code>attribute</code> - the name of the attribute that will make the genes
 unique (e.g. "name", gene_id, etc)</dd><dd><code>secretedProteinsFile</code> - a list of genes that you want to restrict
 your analysis to</dd>
<dt><span class="strong">Returns:</span></dt><dd>the mean intron length of your restricted gene list</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd></dl>
</li>
</ul>
<a name="calculateCodingRegion(org.biojava3.genome.parsers.gff.FeatureList, java.util.HashMap, java.lang.String, double)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>calculateCodingRegion</h4>
<pre>public&nbsp;double&nbsp;calculateCodingRegion(org.biojava3.genome.parsers.gff.FeatureList&nbsp;featList,
                           java.util.HashMap&lt;java.lang.String,int[]&gt;&nbsp;codingMap,
                           java.lang.String&nbsp;attribute,
                           double&nbsp;genomeLength)
                             throws java.io.IOException,
                                    java.lang.Exception</pre>
<div class="block">Calculate the proportion of the genome that contains coding regions (from
 the start of the first exon to the end of the last exon)</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>featList</code> - a org.biojava3.genome.parsers.gff.FeatureList</dd><dd><code>codingMap</code> - a HashMap where the keys are the reference sequence
 names and the values a zero-filled int [] the length of the associated
 reference sequence</dd><dd><code>attribute</code> - the name of the attribute that will make the genes
 unique (e.g. "name", gene_id, etc)</dd><dd><code>genomeLength</code> - the length of the genome</dd>
<dt><span class="strong">Returns:</span></dt><dd>the proportion of the genome which contains exons</dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd>
<dd><code>java.lang.Exception</code></dd></dl>
</li>
</ul>
<a name="getBlocks(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getBlocks</h4>
<pre>public&nbsp;java.util.HashMap&lt;java.lang.String,java.util.ArrayList&lt;java.lang.Integer&gt;&gt;&nbsp;getBlocks(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                                                                                   java.lang.String&nbsp;featureType)</pre>
<div class="block">Creates start/end blocks of features</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fl</code> - a org.biojava3.genome.parsers.gff.FeatureList</dd><dd><code>featureType</code> - the type of feature to make blocks from</dd>
<dt><span class="strong">Returns:</span></dt><dd>a HashMap of blocks where the keys are the names of the reference
 sequences and the values are arrays of Integers with the consecutive
 start and then finish co-ordinates of all the requested featureTypes</dd></dl>
</li>
</ul>
<a name="getBlocks(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, java.lang.String)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getBlocks</h4>
<pre>public&nbsp;java.util.HashMap&lt;java.lang.String,java.util.ArrayList&lt;java.lang.Integer&gt;&gt;&nbsp;getBlocks(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                                                                                   java.lang.String&nbsp;featureType,
                                                                                   java.lang.String&nbsp;attribute)</pre>
<div class="block">Creates start/end blocks of features</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fl</code> - a org.biojava3.genome.parsers.gff.FeatureList</dd><dd><code>featureType</code> - the type of feature to make blocks from</dd><dd><code>attribute</code> - the name of the attribute that will make the genes
 unique (e.g. "name", gene_id, etc)</dd>
<dt><span class="strong">Returns:</span></dt><dd>a HashMap of blocks where the keys are the names of the reference
 sequences and the values are arrays of Integers with the consecutive
 start and then finish co-ordinates of all the requested featureTypes</dd></dl>
</li>
</ul>
<a name="getBlocks(org.biojava3.genome.parsers.gff.FeatureList, java.lang.String, java.lang.String, java.util.Set)">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>getBlocks</h4>
<pre>public&nbsp;java.util.HashMap&lt;java.lang.String,java.util.ArrayList&lt;java.lang.Integer&gt;&gt;&nbsp;getBlocks(org.biojava3.genome.parsers.gff.FeatureList&nbsp;fl,
                                                                                   java.lang.String&nbsp;featureType,
                                                                                   java.lang.String&nbsp;attribute,
                                                                                   java.util.Set&nbsp;secretedProteins)</pre>
<div class="block">Creates start/end blocks of features</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>fl</code> - a org.biojava3.genome.parsers.gff.FeatureList</dd><dd><code>featureType</code> - the type of feature to make blocks from</dd><dd><code>attribute</code> - the name of the attribute that will make the genes
 unique (e.g. "name", gene_id, etc)</dd><dd><code>secretedProteins</code> - a list of genes that you want to restrict your
 analysis to</dd>
<dt><span class="strong">Returns:</span></dt><dd>a HashMap of blocks where the keys are the names of the reference
 sequences and the values are arrays of Integers with the consecutive
 start and then finish co-ordinates of all the requested featureTypes</dd></dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar_bottom">
<!--   -->
</a><a href="#skip-navbar_bottom" title="Skip navigation links"></a><a name="navbar_bottom_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/GFFFeatureStats.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev Class</li>
<li>Next Class</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?uk/ac/tsl/etherington/genomehelper/gff/GFFFeatureStats.html" target="_top">Frames</a></li>
<li><a href="GFFFeatureStats.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
